hw3/03-react/src/views/Search.js

```diff
@@ -0,0 +1,87 @@
+import React, { useState } from "react";
+
+function Search(props) {
+  const { title } = props;
+  const [charCards, setCharCards] = useState([]);
+
+  const addCards = (userInput) => {
+    const getData = async () => {
+      const inputLowerCased = userInput.toLowerCase();
+      const characterData = await fetchData();
+      let count = 0;
+
+      characterData.forEach((character) => {
+        const charLowerCased = character.fullName.toLowerCase();
+        const isMatch = charLowerCased.includes(inputLowerCased);
+
+        if (isMatch && inputLowerCased !== "") {
+          count += 1;
+          const newCard = (
+            <div className="col mt-2 mx-0" key={count}>
+              <div className="pt-2 justify-content-center mx-auto char-card">
+                <img
+                  src={character.imageUrl}
+                  alt={`Portrait of ${character.fullName}`}
+                ></img>
+                <h2 className="mt-3 fw-bold text-center mx-auto w-75 text-white">
+                  {character.fullName}
+                </h2>
+              </div>
+            </div>
+          );
+          setCharCards((previousCards) => [...previousCards, newCard]);
+        }
+      });
+
+      if (count === 0) {
+        const newCard = (
+          <div key="0">
+            <p className="text-danger">No Match Found</p>
+          </div>
+        );
+        setCharCards((previousCards) => [...previousCards, newCard]);
+      }
+    };
+    getData();
+  };
+
+  const clearCards = () => {
+    setCharCards([]);
+  };
+  return (
+    <section className="container w-100 h-75 mt-0 rounded-2 p-5 mx-auto bg-white pt-0 mt-0">
+      <h1 className="mb-4">{title}</h1>
+      <input
+        className="mb-3"
+        aria-label="Search for a Game of Thrones Character by typing their name and hitting enter"
+        onKeyDown={(e) => {
+          if (e.key === "Enter") {
+            const userInput = e.target.value;
+            clearCards();
+            addCards(userInput);
+          }
+        }}
+      ></input>
+      <div className="row" id="imageContainer">
+        {charCards}
+      </div>
+    </section>
+  );
+}
+export default Search;
+
+async function fetchData() {
+  const url = "https://thronesapi.com/api/v2/Characters";
+  try {
+    const response = await fetch(url);
+    const data = await response.json();
+    return data;
+  } catch (error) {
+    console.error("Fetch failed", error);
+  }
+}
```

Comment:
Calls the same API and fetches the same data as that of Houses.js

Fetching the data (calling the exact same API) twice, in both the Houses and Search components is inefficient. The data they both rely on is the same, and additionally both of these components have the same parent (should be something like App.js, not Navbar).

So why not let the data flow from parent to child?

It would make a lot more sense to call the API once in the App component, then pass the data as props to the houses and search components. This would be more efficient since the API would only need to be called once, but the data would be available across the components that need it.

In summary, you can just fetch the data once and then pass it to child components via props.